cmds/core/shasum/shasum.go

```diff
@@ -5,83 +5,82 @@
 package main
 
 import (
+	"bufio"
 	"crypto/sha1"
 	"crypto/sha256"
 	"fmt"
 	"io"
+	"log"
 	"os"
 
 	"github.com/spf13/pflag"
 )
 
+var (
+	algorithm = pflag.IntP("algorithmhttps://github.com/u-root/u-root/pull/2319", "a", 1, "SHA algorithm, valid args are 1 and 256")
+	help      = pflag.BoolP("help", "h", false, "Show this help and exit")
+)
+var usage = "Usage:\nshasum -a <algorithm> <File Name>"
+
 func helpPrinter() {
-	fmt.Printf("Usage:\nshasum -a <algorithm> <File Name>\n")
+	fmt.Println(usage)
 	pflag.PrintDefaults()
-	os.Exit(0)
-}
-
-func versionPrinter() {
-	fmt.Println("shasum utility, URoot Version.")
-	os.Exit(0)
-}
-
-func getInput(fileName string) (input []byte, err error) {
-	if fileName != "" {
-		return os.ReadFile(fileName)
-	}
-	return io.ReadAll(os.Stdin)
 }
 
 //
 // shaPrinter prints sha1/sha256 of given data. The
 // value of algorithm is expected to be 1 for SHA1
 // and 256 for SHA256
 //
-func shaPrinter(algorithm int, data []byte) string {
-	var sha string
-	if algorithm == 256 {
-		sha = fmt.Sprintf("%x", sha256.Sum256(data))
-	} else if algorithm == 1 {
-		sha = fmt.Sprintf("%x", sha1.Sum(data))
+func shaPrinter(w io.Writer, r io.Reader) error {
+	if *algorithm == 256 {
```

Comment:
I would consider a slight difference. 

func shaString(algorithm int) (string, error) {
var h hash.Hash
switch *algorithm {
case 256: h = sha256.New()
case 1: h = sha1.New()
default: return fmt.Errorf("Invalid algorithm, only 1 or 256 are valid") // in addition to telling us what's wrong, tell us what's right!
}
if _, err := io.Copy(h, r); err != nil {
etc.
}
return fmt.Errorf("%#x", h.nil), nil

you can continue to have it print, but you might want to consider this simpler code sructure. I think having it return the string makes tests and debug easier.

Reply:
That is certainly true. On it. Give me 10mins

Reply:
Hmmm i decided against returning the string, but the hash bytes. Someone told me once, we dont want to print too much in functions which do something else....